crates/sui-core/src/consensus_handler.rs

```diff
@@ -703,17 +713,29 @@ impl<C: CheckpointServiceNotify + Send + Sync> ConsensusHandler<C> {
         {
             let span = trace_span!("ConsensusHandler::HandleCommit::process_consensus_txns");
             let _guard = span.enter();
-            for (authority_index, parsed_transactions) in consensus_commit.transactions() {
+            for (block, parsed_transactions) in consensus_commit.transactions() {
+                let author = block.author.value();
                 // TODO: consider only messages within 1~3 rounds of the leader?
-                self.last_consensus_stats
-                    .stats
-                    .inc_num_messages(authority_index as usize);
-                for parsed in parsed_transactions {
-                    // Skip executing rejected transactions. Unlocking is the responsibility of the
-                    // consensus transaction handler.
+                self.last_consensus_stats.stats.inc_num_messages(author);
+                for (tx_index, parsed) in parsed_transactions.into_iter().enumerate() {
+                    let position = ConsensusTxPosition {
+                        block,
+                        index: tx_index as TransactionIndex,
+                    };
                     if parsed.rejected {
+                        // TODO(fastpath): Add metrics for rejected transactions.
+                        if parsed.transaction.kind.is_user_transaction() {
+                            self.epoch_store
+                                .set_consensus_tx_status(position, ConsensusTxStatus::Rejected);
+                        }
+                        // Skip executing rejected transactions. Unlocking is the responsibility of the
+                        // consensus transaction handler.
```

Comment:
```suggestion
                        // Skip executing rejected transactions.
```

Reply:
Actually I think unlocking needs to be done in commit handler, if we get to do it.

Reply:
Potentially yes for reverting votes on GC'ed blocks. But when we get to it, a simpler way might be to GC remaining locks on unfinalized transactions automatically as consensus progresses, similar to consensus tx status cache but persisted.